[MC,ELF] Emit warning if a string constant contains newline char

chestnykh opened the pull request:
GAS emits warning about newline in the string constant so make the same behaviour.

chestnykh committed:
[MC,ELF] Emit warning if a string constant contains newline char.

GAS emits warning about newline in the string constant
so make the same behaviour.

s-barannikov reviewed (dismissed):
1. Printing text to stdout is not the correct way to report warnings, it should be done via source manager (and it is likely there are helper functions).
2. AsmLexer is a lexer, not parser. It does not know the context the lexed string is used. In some contexts the warning may not be desired. The warning, if needed, should be emitted by a parser that knows full context, e.g. in `AsmParser::parseDirectiveAscii`.

chestnykh committed:
[MC,ELF] Move warning handling to parser code

Emit warning about newline characters in strings
for `.string`, '.ascii' and '.asciz' directives
like GAS.

chestnykh commented:
> 1. Printing text to stdout is not the correct way to report warnings, it should be done via source manager (and it is likely there are helper functions).
> 
>     2. AsmLexer is a lexer, not parser. It does not know the context the lexed string is used. In some contexts the warning may not be desired. The warning, if needed, should be emitted by a parser that knows full context, e.g. in `AsmParser::parseDirectiveAscii`.

I reworked

s-barannikov commented on llvm/lib/MC/MCParser/AsmLexer.cpp:
```diff
@@ -646,7 +646,8 @@ AsmToken AsmLexer::LexQuote() {
 
-  // TODO: does gas allow multiline string constants?
+  // gas doesn't allow multiline string constants
+  // and emits a warning if a string constant contains newline character
```
I guess the comment can just be removed.

s-barannikov commented on llvm/test/MC/ELF/warn-newline-in-string-constant.s:
```diff
@@ -0,0 +1,52 @@
+
+// CHECK-WARN:   warn-newline-in-string-constant.s:6:24: warning: unterminated string; newline inserted
+// CHECK-WARN:  .ascii "some test ascii
+// CHECK-WARN:                         ^
```
FileCheck is by default space-insensitive. If you want to test the location of a caret, you need to add `--strict-whitespace` argument.

chestnykh replied:
Done

s-barannikov commented on llvm/lib/MC/MCParser/AsmParser.cpp:
```diff
@@ -3125,6 +3125,16 @@ bool AsmParser::parseDirectiveAscii(StringRef IDVal, bool ZeroTerminated) {
+      // Warn about newline characters in parsed string like GAS.
+      size_t NewlinePos = -1;
+      size_t DataSize = Data.size();
+      const char *Start = getTok().getLoc().getPointer() - DataSize - 1;
```
The location will be incorrect if the string contains escaped characters (e.g. '\xFF'), and I don't see a way how you can reliably get the location of a newline without making `parseEscapedString` return it. Moreover, the string could contain '\n', which must not be diagnosed.

s-barannikov replied:
I think it may make sense to look for newlines and emit warnings inside `parseEscapedString`. If not all callers of the function want these warnings, they can be issues conditionally based on a new boolean argument, but I couldn't find such a caller.

chestnykh replied:
Moved the code to `parseEscapedString`. Seems that all cases are handled correctly, i've tried to write tests to each of them.

s-barannikov commented on llvm/lib/MC/MCParser/AsmParser.cpp:
```diff
@@ -3125,6 +3125,16 @@ bool AsmParser::parseDirectiveAscii(StringRef IDVal, bool ZeroTerminated) {
+      const char *Start = getTok().getLoc().getPointer() - DataSize - 1;
+
+      while ((NewlinePos = Data.find('\n', NewlinePos + 1)) < DataSize)
+        Warning(SMLoc::getFromPointer(Start + NewlinePos),
```
Ideally, you wold want to check what `Warning` returns. It can return `true` in --fatal-warning mode, in which case compilation should fail.

chestnykh replied:
Done

chestnykh committed:
[MC,AsmParser] Warn about newlines in some cases

Move the point of check into `parseEscapedString`.
Check if the `Warning()` returned true to stop compilation.

chestnykh committed:
[MC,AMDGPU] Correct code format

s-barannikov commented on llvm/lib/MC/MCParser/AsmParser.cpp:
```diff
@@ -3123,8 +3129,9 @@ bool AsmParser::parseDirectiveAscii(StringRef IDVal, bool ZeroTerminated) {
     // discusssion at https://reviews.llvm.org/D91460 for more details.
     do {
-      if (parseEscapedString(Data))
+      if (parseEscapedString(Data, /* WarnNewline */ true))
```
The correct style is:
```suggestion
      if (parseEscapedString(Data, /*WarnNewline=*/true))
```

s-barannikov commented on llvm/lib/MC/MCParser/DarwinAsmParser.cpp:
```diff
@@ -613,7 +613,7 @@ bool DarwinAsmParser::parseDirectiveLinkerOption(StringRef IDVal, SMLoc) {
 
     std::string Data;
-    if (getParser().parseEscapedString(Data))
+    if (getParser().parseEscapedString(Data, false))
```
I guess the intent of the warning is to help finding missing quotes in such pieces of assembly:

```asm
.string "asdf
// lots of directives / code here
.string "qwer"
```

The assembler would complain about unknown instruction mnemonic "qwer" because it parsed everything after the very first quote as an argument of the first `.string` directive. This consideration should apply to `.linker_option` directive as well, which is unlikely to expect newlines in the argument. That is, I would prohibit newlines in the argument of most, if not all, directives calling `parseEscapedString`.

chestnykh replied:
So your suggestion is to emit `Error()` instead of `Warning()` for all such directives ?

chestnykh replied:
I agree with you. Such a code with newlines in strange places most likely can only appear if it is written manually.
So i i think that prohibiting of newlines in all `parseEscapedString()` places is a good idea

chestnykh replied:
And if it is just a warning, it will most likely be ignored

chestnykh replied:
> No, I mean the warning should be issued for most/all of such directives, not only for .string/.ascii. Sorry if I wasn't clear.

No problem :)
my understanding was very close.
Okay, then i'm going to remove the second argument of `parseEscapedString` and emit warning unconditionally.

chestnykh replied:
Then i should extend test cases

s-barannikov replied:
A couple of different directives should be enough, other directives do not add new context.

chestnykh replied:
Done

s-barannikov commented on llvm/test/MC/ELF/warn-newline-in-string-constant.s:
```diff
@@ -0,0 +1,64 @@
+
+.string "1234\n\xFF\n\xFF\n"
+
+// CHECK-WARN-NOT: warn-newline-in-string-constant.s:63{{.*}}
```
Missing newline at end of file.

chestnykh replied:
Done

s-barannikov commented on llvm/lib/MC/MCParser/AsmParser.cpp:
```diff
@@ -3021,14 +3021,20 @@ bool AsmParser::parseDirectiveSet(StringRef IDVal, AssignmentKind Kind) {
   for (unsigned i = 0, e = Str.size(); i != e; ++i) {
     if (Str[i] != '\\') {
+      if (Str[i] == '\n' && WarnNewline) {
+        SMLoc NewlineLoc =
+            SMLoc::getFromPointer(getTok().getLoc().getPointer() + i + 1);
```
I think you can reduce this to just the following, but need to check:
```suggestion
        SMLoc NewlineLoc = SMLoc::getFromPointer(Str.data() + i);
```

chestnykh replied:
Yes, it works, thanks

s-barannikov commented on llvm/test/MC/ELF/warn-newline-in-string-constant.s:
```diff
@@ -0,0 +1,64 @@
+// RUN: llvm-mc -filetype=obj -triple x86_64-pc-linux-gnu %s 2>&1 -o %t | FileCheck --strict-whitespace %s --check-prefix=CHECK-WARN
```
Discarded output should go to /dev/null

chestnykh replied:
Done

chestnykh committed:
[MC,AsmParser] Rework `parseEscapedString`

- Remove `WarnNewline` arg
- Emit warning to `\n` symbol unconditionally

chestnykh committed:
[MC,AsmParser] Rework test to newline warnings.

- Add new testcases

chestnykh committed:
[MC,AsmParser] Adjuct test to newline warnings

chestnykh committed:
[MC,AsmParser] Correct code style

MaskRay commented on llvm/lib/MC/MCParser/AsmParser.cpp:
```diff
@@ -3125,6 +3130,7 @@ bool AsmParser::parseDirectiveAscii(StringRef IDVal, bool ZeroTerminated) {
     do {
       if (parseEscapedString(Data))
         return true;
+
```
revert unneeded change

chestnykh replied:
Done

MaskRay commented on llvm/test/MC/ELF/warn-newline-in-escaped-string.s:
```diff
@@ -0,0 +1,69 @@
+
+"
+
+// CHECK-WARN:   warn-newline-in-escaped-string.s:3:21: warning: unterminated string; newline inserted
```
Instead of string/string/string check/check/check, consider check/string check/string check/string that new tests use more.

You can use the following to test relative line numbers, so that inserting lines in the future would not require test update. We often omit the filename.
```
[[#@LINE+1]]:21: warning: 
```

chestnykh replied:
Thx, reworked

chestnykh committed:
[MC,AsmParser] Refactor AsmParser.c and the test

- Revert unneeded change
- Refactor test to use string/check style and
[[#@LINE]] instead of hardcoded line numbers

chestnykh committed:
[MC,AsmParser] Remove unneeded filename from the test

MaskRay commented on llvm/test/MC/ELF/warn-newline-in-escaped-string.s:
```diff
@@ -0,0 +1,69 @@
+// CHECK-WARN:   [[#@LINE-8]]:28: warning: unterminated string; newline inserted
+// CHECK-WARN:   .asciz "another test string
+// CHECK-WARN:   [[#@LINE-9]]:1: warning: unterminated string; newline inserted
+// CHECK-WARN:   ^
```
You can delete `^` checks. With column numbers on the above line, `^` check isn't useful.

Use `-NEXT` whenever applicable

chestnykh replied:
Done

MaskRay commented on llvm/test/MC/ELF/warn-newline-in-escaped-string.s:
```diff
@@ -0,0 +1,69 @@
+// CHECK-WARN:   [[#@LINE-2]]:20: warning: unterminated string; newline inserted
+
+.ascii "test\nstring\xFF\n\n\xFF"
+// CHECK-WARN-NOT:    [[#@LINE-1]]{{.*}}
```
You can use `FileCheck --implicit-check-not=warning:` and remove these `-NOT:` patterns

chestnykh replied:
Done

chestnykh committed:
[MC,AsmParser] Refactor test

- Use `CHECK-NEXT`
- Use `implicit-check-not` instead of `CHECK-WARN-NOT`

MaskRay commented on llvm/test/MC/ELF/warn-newline-in-escaped-string.s:
```diff
@@ -0,0 +1,50 @@
+// RUN: llvm-mc -filetype=obj -triple x86_64-pc-linux-gnu %s 2>&1 -o /dev/null | FileCheck -DFILE=%s --strict-whitespace %s --implicit-check-not=valid1_string --implicit-check-not=valid2_string --implicit-check-not=valid3_string --check-prefix=CHECK-WARN
```
Just use x86_64 as the triple to emphasize that the behavior is for all ELF (actually other object file formats as well, but we don't bother testing), not just Linux.

chestnykh replied:
Done. I don't have write access so please merge this PR when it is ready

chestnykh committed:
[MC] Use x86_64 as triple in test

s-barannikov commented on llvm/test/MC/ELF/warn-newline-in-escaped-string.s:
```diff
@@ -0,0 +1,50 @@
+// RUN: llvm-mc -filetype=obj -triple x86_64 %s 2>&1 -o /dev/null | FileCheck -DFILE=%s --strict-whitespace %s --implicit-check-not=valid1_string --implicit-check-not=valid2_string --implicit-check-not=valid3_string --check-prefix=CHECK-WARN
```
There should be just one --implicit-check-not with a regex for the undesired warning. This will work as long as all expected warnings are checked by CHECK lines.

chestnykh replied:
like `--implicit-check-not=valid*_string` ?

s-barannikov replied:
E.g. `--implicit-check-not="{{[0-9]+:[0-9]+: warning: unterminated string}}"`

chestnykh replied:
Thanks, got it

chestnykh replied:
Done

chestnykh committed:
[MC, ELF] Refactor test

- Add one `--implicit-check-not` with the appropriate pattern
- Refactor checks

s-barannikov commented on llvm/test/MC/ELF/warn-newline-in-escaped-string.s:
```diff
@@ -0,0 +1,54 @@
+"
+
+// CHECK-WARN:   [[#@LINE-7]]:24: warning: unterminated string; newline inserted
+// CHECK-NEXT:   .ascii "some test ascii
```
These CHECK-NEXT directives check nothing because you explicitly renamed the default "CHECK" to "CHECK-WARN" by --check-prefix option. The correct directive would therefore be "CHECK-WARN-NEXT". 

--check-prefix is only useful when you have multiple RUN lines. Using it when you have single RUN line does not hurt, but is just redundant.

(I'm not sure it is worth checking what these -NEXT directives are supposed to check. I'd just remove them, but you can keep them if you like, just fix the spelling of the directives.)

chestnykh replied:
Done

s-barannikov commented on llvm/test/MC/ELF/warn-newline-in-escaped-string.s:
```diff
@@ -0,0 +1,54 @@
+// RUN: llvm-mc -filetype=obj -triple x86_64 %s 2>&1 -o /dev/null | FileCheck -DFILE=%s --strict-whitespace %s --implicit-check-not="{{[0-9]+:[0-9]+: warning: unterminated string}}" --check-prefix=CHECK-WARN
```
```suggestion
// RUN: llvm-mc -filetype=obj -triple x86_64 %s 2>&1 -o /dev/null \
//   | FileCheck %s --implicit-check-not="{{[0-9]+:[0-9]+: warning: unterminated string}}" --check-prefix=CHECK-WARN
```

-DFILE is unused and --strict-whitespace no longer seems necessary (?)
Long RUN lines can be split into multiple by the backslash character.

chestnykh replied:
Done

chestnykh committed:
[MC,ELF] Refactor the test

- Remove redundant `--check-prefix`
- Remove `-DFILE=`

s-barannikov reviewed (approved):
LGTM

MaskRay commented on llvm/test/MC/ELF/warn-newline-in-escaped-string.s:
```diff
@@ -0,0 +1,55 @@
+// RUN: llvm-mc -filetype=obj -triple x86_64 %s 2>&1 -o /dev/null \
+// RUN: | FileCheck %s --implicit-check-not="{{[0-9]+:[0-9]+: warning: unterminated string}}"
```
We usually just test `--implicit-check-not=warning:`

chestnykh replied:
Done

s-barannikov commented:
I've rebased the branch to check if that would resolve CI failure (it looked unrelated).

chestnykh committed:
[MC,ELF] Simplify `--implicit-check-not` argument

chestnykh committed:
[MC,ELF] Add missed colon

MaskRay commented on llvm/test/MC/ELF/warn-newline-in-escaped-string.s:
```diff
@@ -0,0 +1,55 @@
+// RUN: llvm-mc -filetype=obj -triple x86_64 %s 2>&1 -o /dev/null \
+// RUN: | FileCheck %s --implicit-check-not=warning:
```
continuation lines should be indented by 2

chestnykh replied:
Done

chestnykh committed:
[MC,ELF] Add needed indent
